File: tools/gen_units.py
#!/usr/bin/env python3
"""Generate src/intl_units_data.zig from CLDR en units.json."""
import json, sys

d = json.load(open('/tmp/cldr_units.json'))['main']['en']['units']

def suffix(pat):
    # pattern is "{0} meter" / "{0}m" / "{0}°C"; return the part after {0}.
    i = pat.find('{0}')
    return pat[i+3:] if i >= 0 else pat

# Build per-width maps: JS unit id -> (one_suffix, other_suffix)
widths = ['long', 'short', 'narrow']
units = {}   # id -> {width: (one, other)}
per_suffix = {}  # id -> {width: perUnitPattern-suffix} (denominator form, e.g. "/s")
compound = {}    # width -> compoundUnitPattern "{0}/{1}"

for w in widths:
    comp = d[w].get('per', {}).get('compoundUnitPattern')
    if comp:
        compound[w] = comp
    for key, val in d[w].items():
        if not isinstance(val, dict):
            continue
        if '-' not in key:
            continue
        uid = key.split('-', 1)[1]  # strip category (length-meter -> meter)
        one = val.get('unitPattern-count-one')
        other = val.get('unitPattern-count-other')
        if one is None and other is None:
            continue
        one = one if one is not None else other
        other = other if other is not None else one
        units.setdefault(uid, {})
        if w not in units[uid]:
            units[uid][w] = (suffix(one), suffix(other))
        pu = val.get('perUnitPattern')
        if pu is not None:
            per_suffix.setdefault(uid, {})[w] = suffix(pu)

def zesc(s):
    return s.replace('\\', '\\\\').replace('"', '\\"')

# Emit a table: unit id + 6 suffixes (l1,lo,s1,so,n1,no) + per denom forms (pl,ps,pn)
lines = []
lines.append("//! GENERATED from CLDR en units.json (see tools/gen_units.py). English")
lines.append("//! unit patterns for Intl.NumberFormat style:\"unit\": the text that follows")
lines.append("//! the number, per width (long/short/narrow) and plural (one/other), plus")
lines.append("//! the per-unit denominator suffix for compound X-per-Y composition.")
lines.append("")
lines.append("pub const Unit = struct {")
lines.append("    id: []const u8,")
lines.append("    l1: []const u8, lo: []const u8, // long one/other")
lines.append("    s1: []const u8, so: []const u8, // short one/other")
lines.append("    n1: []const u8, no: []const u8, // narrow one/other")
lines.append("    pl: []const u8, ps: []const u8, pn: []const u8, // per-suffix long/short/narrow (denominator)")
lines.append("};")
lines.append("")
lines.append("pub const units = [_]Unit{")
for uid in sorted(units):
    u = units[uid]
    def g(w, i):
        return zesc(u.get(w, ('', ''))[i])
    ps = per_suffix.get(uid, {})
    lines.append('    .{ .id = "%s", .l1 = "%s", .lo = "%s", .s1 = "%s", .so = "%s", .n1 = "%s", .no = "%s", .pl = "%s", .ps = "%s", .pn = "%s" },' % (
        zesc(uid), g('long',0), g('long',1), g('short',0), g('short',1), g('narrow',0), g('narrow',1),
        zesc(ps.get('long','')), zesc(ps.get('short','')), zesc(ps.get('narrow',''))))
lines.append("};")
lines.append("")
# compound patterns
lines.append('pub const compound_long = "%s";' % zesc(compound.get('long','{0}/{1}')))
lines.append('pub const compound_short = "%s";' % zesc(compound.get('short','{0}/{1}')))
lines.append('pub const compound_narrow = "%s";' % zesc(compound.get('narrow','{0}/{1}')))
lines.append("")
sys.stdout.write("\n".join(lines))
sys.stderr.write("units=%d compound=%s\n" % (len(units), compound))

File: tools/gen_units.ts
/** Generate src/intl_units_data.zig from CLDR English units.json. */
import { readText } from "./lib/home";

const data = JSON.parse(readText(process.argv[2] || "/tmp/cldr_units.json")).main.en.units;
const widths = ["long", "short", "narrow"];
const suffix = (pattern: string) => { const index = pattern.indexOf("{0}"); return index >= 0 ? pattern.slice(index + 3) : pattern; };
const escapeZig = (text: string) => text.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
const units: Record<string, Record<string, string[]>> = {};
const perSuffix: Record<string, Record<string, string>> = {};
const compound: Record<string, string> = {};
for (const width of widths) {
  if (data[width].per && data[width].per.compoundUnitPattern) compound[width] = data[width].per.compoundUnitPattern;
  for (const key of Object.keys(data[width])) {
    const value = data[width][key];
    if (!value || typeof value !== "object" || key.indexOf("-") < 0) continue;
    const id = key.slice(key.indexOf("-") + 1);
    let one = value["unitPattern-count-one"];
    let other = value["unitPattern-count-other"];
    if (one == null && other == null) continue;
    one = one == null ? other : one;
    other = other == null ? one : other;
    if (!units[id]) units[id] = {};
    if (!units[id][width]) units[id][width] = [suffix(one), suffix(other)];
    if (value.perUnitPattern != null) {
      if (!perSuffix[id]) perSuffix[id] = {};
      perSuffix[id][width] = suffix(value.perUnitPattern);
    }
  }
}
const lines = [
  "//! GENERATED from CLDR en units.json (see tools/gen_units.ts). English",
  "//! unit patterns for Intl.NumberFormat style:\"unit\": the text that follows",
  "//! the number, per width (long/short/narrow) and plural (one/other), plus",
  "//! the per-unit denominator suffix for compound X-per-Y composition.", "",
  "pub const Unit = struct {", "    id: []const u8,",
  "    l1: []const u8, lo: []const u8, // long one/other",
  "    s1: []const u8, so: []const u8, // short one/other",
  "    n1: []const u8, no: []const u8, // narrow one/other",
  "    pl: []const u8, ps: []const u8, pn: []const u8, // per-suffix long/short/narrow (denominator)",
  "};", "", "pub const units = [_]Unit{",
];
for (const id of Object.keys(units).sort()) {
  const get = (width: string, index: number) => escapeZig((units[id][width] || ["", ""])[index]);
  const per = perSuffix[id] || {};
  lines.push(`    .{ .id = "${escapeZig(id)}", .l1 = "${get("long", 0)}", .lo = "${get("long", 1)}", .s1 = "${get("short", 0)}", .so = "${get("short", 1)}", .n1 = "${get("narrow", 0)}", .no = "${get("narrow", 1)}", .pl = "${escapeZig(per.long || "")}", .ps = "${escapeZig(per.short || "")}", .pn = "${escapeZig(per.narrow || "")}" },`);
}
lines.push("};", "", `pub const compound_long = "${escapeZig(compound.long || "{0}/{1}")}";`, `pub const compound_short = "${escapeZig(compound.short || "{0}/{1}")}";`, `pub const compound_narrow = "${escapeZig(compound.narrow || "{0}/{1}")}";`, "");
process.stdout.write(lines.join("\n"));
const compoundRepr = "{" + widths.filter(width => compound[width] != null).map(width => `'${width}': '${compound[width]}'`).join(", ") + "}";
process.stderr.write(`units=${Object.keys(units).length} compound=${compoundRepr}\n`);
